Content.Server/Nyanotrasen/Research/Oracle/OracleSystem.cs

```diff
@@ -234,13 +237,33 @@ public List<EntProtoId> GetAllProtos(OracleComponent component)
         var allTechs = _prototypeManager.EnumeratePrototypes<TechnologyPrototype>();
         var allRecipes = new List<EntProtoId>();
 
+        var researchServers = new List<(
+            EntityUid serverUid, 
+            TechnologyDatabaseComponent database, 
+            Dictionary<string, int> disciplineTiers
+        )>();
```

Comment:
Use `Entity<TComponent>` instead of separate `EntityUid` and `database`
```suggestion
        var researchServers = new List<(Entity<TechnologyDatabaseComponent> database, Dictionary<string, int> disciplineTiers)>();
```

Reply:
I'm curious, by the way. Do you know what the point of this Resolve check is? That's the only reason I keep the uid

`SharedResearchSystem.cs` line 215
```
    public bool IsTechnologyUnlocked(EntityUid uid, string technologyId, TechnologyDatabaseComponent? component = null)
    {
        return Resolve(uid, ref component, false) && component.UnlockedTechnologies.Contains(technologyId);
    }
```

Is this just like, checking that the server exists? Could you ever have just a detached TechnologyDatabaseComponent on accident?

Reply:
The component parameter is optional
And it ensures that the component belongs to that entity